exonum-java-proofs/src/main/java/com/exonum/binding/hash/AbstractNonStreamingHashFunction.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2011 The Guava Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+/*
+ * Copyright 2018 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.hash;
+
+import com.google.common.base.Preconditions;
+import com.google.errorprone.annotations.Immutable;
+import java.io.ByteArrayOutputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+
+/**
+ * Skeleton implementation of {@link HashFunction}, appropriate for non-streaming algorithms. All
+ * the hash computation done using {@linkplain #newHasher()} are delegated to the {@linkplain
+ * #hashBytes(byte[], int, int)} method.
+ *
+ * @author Dimitris Andreou
+ */
+@Immutable
+abstract class AbstractNonStreamingHashFunction extends AbstractHashFunction {
```

Comment:
Is this (and the second below) class required to test some of the functionality we already have?

Reply:
Yes, we need that for `AbstractStreamingHasher`, which in turn we need for `FunnelsTest`.

Reply:
I see, do you think it makes sense to put these files in `test/` folder, or keep here in `main/` to be consistent with the source of this package?

Reply:
I think we should keep those in main for consistency with original Guava lib. Plus someone might want to use those outside of tests.

Reply:
OK, the latter is unlikely for they are package-private.